src/xrpld/app/tx/detail/MPTokenIssuanceSet.cpp

```diff
@@ -0,0 +1,118 @@
+  ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+  OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#include <xrpld/app/tx/detail/MPTokenIssuanceSet.h>
+#include <xrpld/ledger/View.h>
+#include <xrpl/protocol/Feature.h>
+#include <xrpl/protocol/TxFlags.h>
+#include <xrpl/protocol/st.h>
+
+namespace ripple {
+
+NotTEC
+MPTokenIssuanceSet::preflight(PreflightContext const& ctx)
+{
+    if (!ctx.rules.enabled(featureMPTokensV1))
+        return temDISABLED;
+
+    if (auto const ret = preflight1(ctx); !isTesSuccess(ret))
+        return ret;
+
+    auto const txFlags = ctx.tx.getFlags();
+
+    // check flags
+    if (txFlags & tfMPTokenIssuanceSetMask)
+        return temINVALID_FLAG;
+    // fails if both flags are set
+    else if ((txFlags & tfMPTLock) && (txFlags & tfMPTUnlock))
+        return temINVALID_FLAG;
+
+    auto const accountID = ctx.tx[sfAccount];
+    auto const holderID = ctx.tx[~sfMPTokenHolder];
+    if (holderID && accountID == holderID)
+        return temMALFORMED;
+
+    return preflight2(ctx);
+}
+
+TER
+MPTokenIssuanceSet::preclaim(PreclaimContext const& ctx)
+{
+    // ensure that issuance exists
+    auto const sleMptIssuance =
+        ctx.view.read(keylet::mptIssuance(ctx.tx[sfMPTokenIssuanceID]));
+    if (!sleMptIssuance)
+        return tecOBJECT_NOT_FOUND;
+
+    // if the mpt has disabled locking
+    if (!((*sleMptIssuance)[sfFlags] & lsfMPTCanLock))
+        return tecNO_PERMISSION;
+
+    // ensure it is issued by the tx submitter
+    if ((*sleMptIssuance)[sfIssuer] != ctx.tx[sfAccount])
+        return tecNO_PERMISSION;
+
+    if (auto const holderID = ctx.tx[~sfMPTokenHolder])
+    {
+        // make sure holder account exists
+        if (!ctx.view.exists(keylet::account(*holderID)))
+            return tecNO_DST;
+
+        // the mptoken must exist
+        if (!ctx.view.exists(
+                keylet::mptoken(ctx.tx[sfMPTokenIssuanceID], *holderID)))
+            return tecOBJECT_NOT_FOUND;
+    }
+
+    return tesSUCCESS;
+}
+
+TER
+MPTokenIssuanceSet::doApply()
+{
+    auto const mptIssuanceID = ctx_.tx[sfMPTokenIssuanceID];
+    auto const txFlags = ctx_.tx.getFlags();
+    auto const holderID = ctx_.tx[~sfMPTokenHolder];
+    std::shared_ptr<SLE> sle;
+
+    if (holderID)
+        sle = view().peek(keylet::mptoken(mptIssuanceID, *holderID));
+    else
+        sle = view().peek(keylet::mptIssuance(mptIssuanceID));
+
+    if (!sle)
+        return tecINTERNAL;
+
+    std::uint32_t const flagsIn = sle->getFieldU32(sfFlags);
+    std::uint32_t flagsOut = flagsIn;
+
+    if (txFlags & tfMPTLock)
+        flagsOut |= lsfMPTLocked;
+    else if (txFlags & tfMPTUnlock)
+        flagsOut &= ~lsfMPTLocked;
+
+    if (flagsIn != flagsOut)
+        sle->setFieldU32(sfFlags, flagsOut);
+
+    view().update(sle);
```

Comment:
I think we'd like to keep it here because other transactors have it similar (eg. AccountSet). Also potential in the future we may add new functionalities to this transactor so might as well leave it here.